Adding support to SQL cmdlets to allow provisioning of ResourceIdentity at create or update time for SQL server resources.

mihymel opened the pull request:
Adding support to SQL cmdlets to allow provisioning of ResourceIdentity at create or update time for SQL server resources.

<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
This includes changes to SQL cmdlets "New-AzureRmSqlServer" and "Set-AzureRmSqlServer" to add an optional "AssignIdentity" switch to support provisioning of an AAD identity for the server.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

shahabhijeet commented:
@mihymel please provide description for your PR.
We have provided a template, please use it as guideline to provide information about your PR.
Helps in understanding the PR much faster/better.

cormacpayne commented:
@akromm Hey Adam, would you mind taking a look at this PR when you get the chance?

akromm commented on src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/Common.ps1:
```diff
@@ -223,43 +223,19 @@ Gets the values of the parameters used in the Server Key Vault Key tests
+			  keyName = "ps-key";
+			  location = "westeurope";
+			  keyId = "https://ps-vault.vault.azure.net/keys/ps-key/65011ae18b5747eba29ea8e5fc85c1a9";
+			  serverKeyName = "ps-vault_ps-key_65011ae18b5747eba29ea8e5fc85c1a9";
```
None of these parameters should be hard coded. It makes it impossible to re-generate the tests. The tests should be recordable by anyone.

mihymel replied:
I removed any changes relating to updating the ServerKey and EncryptionProtector related cmdlets to the Swagger generated SDK, so that this change is only related to Identity changes for Server.

akromm commented on src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ServerCrudTests.ps1:
```diff
@@ -139,3 +139,64 @@ function Test-RemoveServer
+		# With all parameters
+		$server1 = New-AzureRmSqlServer -ResourceGroupName $rg.ResourceGroupName -ServerName $serverName -Location "northeurope" -ServerVersion $version -SqlAdministratorCredentials $credentials -AssignIdentity
+		Assert-AreEqual $server1.ServerName $serverName
+		Assert-AreEqual $server1.ServerVersion $version
```
This check is unnecessary.

akromm commented on src/ResourceManager/Sql/Commands.Sql/Common/ResourceIdentityHelper.cs:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    
```
Nit: remove blank lines

akromm commented on src/ResourceManager/Sql/Commands.Sql/ServerKeyVaultKey/Services/AzureSqlServerKeyVaultKeyAdapter.cs:
```diff
@@ -84,13 +84,10 @@ public IList<AzureSqlServerKeyVaultKeyModel> List(string resourceGroupName, stri
         public AzureSqlServerKeyVaultKeyModel CreateOrUpdate(AzureSqlServerKeyVaultKeyModel model)
         {
-            var resp = Communicator.CreateOrUpdate(model.ResourceGroupName, model.ServerName, model.ServerKeyName, new ServerKeyCreateOrUpdateParameters()
+            var resp = Communicator.CreateOrUpdate(model.ResourceGroupName, model.ServerName, model.ServerKeyName, new ServerKey()
```
This is a breaking change, so make sure to increment the version number appropriately

mihymel replied:
I removed any changes relating to updating the ServerKey and EncryptionProtector related cmdlets to the Swagger generated SDK, so that this change is only related to Identity changes for Server.

akromm commented on src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ServerCrudTests.ps1:
```diff
@@ -139,3 +139,94 @@ function Test-RemoveServer
+	$rg = Create-ResourceGroupForTest
+	 	
+	$serverName = Get-ServerName
+	$version = "12.0"
```
This is not used anywhere

akromm commented on src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ServerCrudTests.ps1:
```diff
@@ -139,3 +139,94 @@ function Test-RemoveServer
+	$rg = Create-ResourceGroupForTest
+	 	
+	$serverName = Get-ServerName
+	$version = "12.0"
```
Shouldn't be needed.

mihymel commented:
@shahabhijeet Apologies for the empty title and description, I made a mistake when initially creating the PR. I've updated the title and description and added the "needs-review" label.

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/help/New-AzureRmSqlServer.md:
````diff
@@ -39,6 +39,19 @@ This command creates a version 12 Azure SQL Database server.
 ## PARAMETERS
 
+### -AssignIdentity
+Generate and assign an Azure Active Directory Identity for this server for use with key management services like Azure KeyVault.```yaml
````
@mihymel the `` ```yaml `` at the end of this line should be on a new line by itself

mihymel replied:
@cormacpayne Weird. I used platyPS to generate these. I can fix this, but is this a common issue with platyPS generated help mark down?

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/help/Set-AzureRmSqlServer.md:
````diff
@@ -40,6 +41,19 @@ This command resets the administrator password on the AzureSQL Server named serv
 ## PARAMETERS
 
+### -AssignIdentity
+Generate and assign an Azure Active Directory Identity for this server for use with key management services like Azure KeyVault.```yaml
````
@mihymel same comment about the `` ```yaml `` text

cormacpayne commented:
On-demand: http://azuresdkci.cloudapp.net/job/powershell-demand/1704/

mihymel commented:
@cormacpayne I accidentally dismissed your review after pushing a change to address your comments. Can you review/sign off again? Sorry about this!
